// ngx-youtube-player/esm5/lib/ngx-youtube-player.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { YoutubePlayerComponent } from './ngx-youtube-player.component';
import { YoutubePlayerService } from './ngx-youtube-player.service';
var NgxYoutubePlayerModule = /** @class */ (function () {
    function NgxYoutubePlayerModule() {
    }
    /**
     * @return {?}
     */
    NgxYoutubePlayerModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: NgxYoutubePlayerModule,
            providers: [YoutubePlayerService]
        };
    };
    NgxYoutubePlayerModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [YoutubePlayerComponent],
                    imports: [],
                    providers: [YoutubePlayerService],
                    exports: [YoutubePlayerComponent]
                },] }
    ];
    return NgxYoutubePlayerModule;
}());
export { NgxYoutubePlayerModule };
